pkg/cloud/affinity_groups_test.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloud_test
+
+import (
+	"errors"
+
+	"github.com/apache/cloudstack-go/v2/cloudstack"
+	infrav1 "github.com/aws/cluster-api-provider-cloudstack/api/v1beta1"
+	"github.com/aws/cluster-api-provider-cloudstack/pkg/cloud"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+var _ = Describe("AffinityGroup Unit Tests", func() {
+	var ( // Declare shared vars.
+		mockCtrl    *gomock.Controller
+		mockClient  *cloudstack.CloudStackClient
+		ags         *cloudstack.MockAffinityGroupServiceIface
+		fakeAG      *cloud.AffinityGroup
+		cluster     *infrav1.CloudStackCluster
+		machine     *infrav1.CloudStackMachine
+		capiMachine *capiv1.Machine
+		client      cloud.Client
+	)
+
+	BeforeEach(func() {
+		// Setup new mock services.
+		mockCtrl = gomock.NewController(GinkgoT())
+		mockClient = cloudstack.NewMockClient(mockCtrl)
+		ags = mockClient.AffinityGroup.(*cloudstack.MockAffinityGroupServiceIface)
+		client = cloud.NewClientFromCSAPIClient(mockClient)
+		fakeAG = &cloud.AffinityGroup{
+			Name: "FakeAffinityGroup",
+			Type: cloud.AffinityGroupType}
+		cluster = &infrav1.CloudStackCluster{Spec: infrav1.CloudStackClusterSpec{
+			Zone: "Zone1", Network: "SharedGuestNet1"}}
+		machine = &infrav1.CloudStackMachine{Spec: infrav1.CloudStackMachineSpec{
+			Offering: "Medium Instance", Template: "Ubuntu20"}}
+		machine.ObjectMeta.SetName("rejoshed-affinity-group-test-vm")
+		capiMachine = &capiv1.Machine{}
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+	})
+
+	It("fetches an affinity group", func() {
+		ags.EXPECT().GetAffinityGroupByName(fakeAG.Name).Return(&cloudstack.AffinityGroup{}, 1, nil)
```

Comment:
These cover happy day GetOrCreateAffinityGroup. What's the plan for the other functions in affinity_groups.go?

Reply:
(Or do the integration tests below effectively cover the rest?)